pkg/auth/rbac_test.go

```diff
@@ -87,3 +90,74 @@ func TestValidateRbacPermission(t *testing.T) {
 		})
 	}
 }
+
+func TestCheckUserPermissions(t *testing.T) {
+	tcs := []struct {
+		Name        string
+		rbacConfig  RBACConfig
+		user        *User
+		env         string
+		envGroup    string
+		application string
+		action      string
+		WantError   string
+	}{
+		{
+			Name:        "Check user permission works as expected",
+			user:        &User{DexAuthContext: &DexAuthContext{Role: "Developer"}},
+			env:         "production",
+			envGroup:    "production",
+			application: "app1",
+			action:      PermissionCreateLock,
+			rbacConfig: RBACConfig{DexEnabled: true, Policy: map[string]*Permission{
+				"Developer,CreateLock,production:production,app1,allow": {Role: "Developer"},
+			}},
+		},
+		{
+			Name:        "Application Wildcard works as expected",
+			user:        &User{DexAuthContext: &DexAuthContext{Role: "Developer"}},
+			env:         "production",
+			envGroup:    "production",
+			application: "app1",
+			action:      PermissionCreateLock,
+			rbacConfig: RBACConfig{DexEnabled: true, Policy: map[string]*Permission{
+				"Developer,CreateLock,production:production,*,allow": {Role: "Developer"},
+			}},
+		},
+		{
+			Name:        "Environment independent works as expected",
+			user:        &User{DexAuthContext: &DexAuthContext{Role: "Developer"}},
+			env:         "production",
+			envGroup:    "production",
+			application: "app1",
+			action:      PermissionCreateUndeploy,
+			rbacConfig: RBACConfig{DexEnabled: true, Policy: map[string]*Permission{
+				"Developer,CreateUndeploy,production:*,*,allow": {Role: "Developer"},
+			}},
+		},
+		{
+			Name:        "User does not have permission",
+			user:        &User{DexAuthContext: &DexAuthContext{Role: "Developer"}},
+			env:         "production",
+			envGroup:    "production",
+			application: "app1",
+			action:      PermissionCreateLock,
+			rbacConfig: RBACConfig{DexEnabled: true, Policy: map[string]*Permission{
+				"Developer,CreateLock,staging:staging,app1,allow": {Role: "Developer"},
+			}},
+			WantError: status.Errorf(codes.PermissionDenied, fmt.Sprintf("user does not have permissions for: %s", "Developer,CreateLock,production:production,app1,allow")).Error(),
+		},
+	}
+
+	for _, tc := range tcs {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			err := CheckUserPermissions(tc.rbacConfig, tc.user, tc.env, tc.envGroup, tc.application, tc.action)
+			if err != nil {
```

Comment:
please also check the case `err = nil && wantError != nil`

Reply:
That's not what I intended.
The test must be aware of all cases, e.g. `err = nil && wantError != ""` but also `err != nil && wantError != ""` and err != nil && wantError == ""

Reply:
Ah true... Sorry misunderstood. Instead of that I added `cmpopts.EquateErrors()` that covers all cases mentioned.